keras/utils/tree.py

```diff
@@ -1,60 +1,331 @@
-import tree
+import collections
+import collections.abc
+import types
 
+import optree
 
+from keras.api_export import keras_export
+from keras.backend.config import backend
+
+try:
+    import wrapt
+
+    ObjectProxy = wrapt.ObjectProxy
+except ImportError:
+
+    class ObjectProxy(object):
+        """Stub-class for `wrapt.ObjectProxy``."""
+
+
+# Register backend-specific node classes
+if backend() == "tensorflow":
+    from tensorflow.python.trackable.data_structures import ListWrapper
+
+    optree.register_pytree_node(
+        ListWrapper,
+        lambda x: (x, None),
+        lambda metadata, children: ListWrapper(list(children)),
+        namespace="keras",
+    )
+
+
+@keras_export("keras.utils.tree.is_nested")
```

Comment:
What is the case for exporting these utils to the public API?

Reply:
Kindly ping @mattdangerw as mentioned in https://github.com/keras-team/keras/issues/18614#issuecomment-1764960732
> If we add a `keras.utils.tree` let's make sure to export it.

I have confirmed that:
keras_cv requires `tree.flatten` and `tree.map_structure` (https://github.com/search?q=repo%3Akeras-team%2Fkeras-cv%20tree.&type=code)
keras_nlp also requires `tree.flatten` and `tree.map_strcture` (https://github.com/search?q=repo%3Akeras-team%2Fkeras-nlp%20tree.&type=code)

Exporting these 2 apis may be sufficient?

Reply:
If we have these 2 we might as well export everything. I'd recommend moving them to `keras.tree`.

Reply:
Done. Exports APIs using `keras.tree.*`